### DISCUSSION TITLE
Asynchronous tasks inside a Ghidra Extension

### DISCUSSION THREAD

**Discussion by kn0wl3dge** [NationalSecurityAgency/ghidra#3381] [Q&A] [answered]
Hi,

I am developing a Ghidra Extension that query an API with the current program sha256 and then fill a table with the results.
So I just have a ProgramPlugin, a ComponentProvider and a TableModel.

The thing is the API queries can take more than 30sec (this is intended) and Ghidra completely freezes during this period of time.

What would be the best way to make asynchronous tasks inside a Ghidra Extension ?

**Comment by ryanmkurtz** [comment, marked as answer]
I can only give a quick answer right now...Ghidra provides a Task API for this sort of thing. Try searching for classes that start with Task for examples.

**Reply by kn0wl3dge** [reply]
Thanks for your quick answer! 
Here is an example code in case someone is having the same question:
```java
Task t = new Task("Task Title", true, false, false) {
	@Override
	public void run(TaskMonitor monitor) {
		// Asynchronous code
	}
};
new TaskLauncher(t, tool.getToolFrame(), 0);
```

**Reply by astrelsky** [reply]
> Thanks for your quick answer!
> Here is an example code in case someone is having the same question:
> 
> ```java
> Task t = new Task("Task Title", true, false, false) {
> 	@Override
> 	public void run(TaskMonitor monitor) {
> 		// Asynchronous code
> 	}
> };
> new TaskLauncher(t, tool.getToolFrame(), 0);
> ```

When running in a `GhidraScript` or when you have access to a `PluginTool` you can just use `PluginTool::execute` to start your task.
```java
Task t = new Task("Task Title", true, false, false) {
	@Override
	public void run(TaskMonitor monitor) {
		// Asynchronous code
	}
};
state.getTool().execute(t, delay); // delay is optional
```

**Reply by dev747368** [reply]
My fav:
```java
TaskLauncher.launchModal("Task Title", monitor -> doSlowStuffHere(monitor) );
```

**Reply by astrelsky** [reply]
> My fav:
> 
> ```java
> TaskLauncher.launchModal("Task Title", monitor -> doSlowStuffHere(monitor) );
> ```

`MonitoredRunnable` is not a `FunctionalInterface`.

You can do this though:
```java
TaskLauncher.launchModal("Task Title", () -> doSlowStuffHere());
```

**Reply by dev747368** [reply]
Not sure what your point was re: FunctionalInterface, but:

https://github.com/NationalSecurityAgency/ghidra/blob/91aea06c98c63e590f80e20c677507f769605539/Ghidra/Features/GraphServices/src/main/java/ghidra/graph/export/GraphExporterDialog.java#L265-L266

**Reply by astrelsky** [reply]
> Not sure what your point was re: FunctionalInterface, but:
> 
> https://github.com/NationalSecurityAgency/ghidra/blob/91aea06c98c63e590f80e20c677507f769605539/Ghidra/Features/GraphServices/src/main/java/ghidra/graph/export/GraphExporterDialog.java#L265-L266

A lambda requires a `FunctionalInterface` or it won't compile. Mostly a gripe that I have to make a class a lot of the time when I should be able to just use a lambda.

**Reply by dev747368** [reply]
Hopefully we aren't talking past each other, but:

[Java functional interface doc](https://docs.oracle.com/javase/8/docs/api/java/lang/FunctionalInterface.html)
>However, the compiler will treat any interface meeting the definition of a functional interface as a functional interface regardless of whether or not a FunctionalInterface annotation is present on the interface declaration.

But you are correct that there are a lot of situations where something is almost a candidate, but you can't use a lambda.  Especially some of the older awt / swing listener APIs!  So frustrating.

**Reply by topia** [reply]
I'm not sure exactly why/how for now, but it can compile with generating `MonitorRunnable` (without `FunctionalInterface`).
https://github.com/NationalSecurityAgency/ghidra/blob/91aea06c98c63e590f80e20c677507f769605539/Ghidra/Framework/Docking/src/main/java/ghidra/util/task/TaskLauncher.java#L112
https://github.com/NationalSecurityAgency/ghidra/blob/91aea06c98c63e590f80e20c677507f769605539/Ghidra/Framework/Generic/src/main/java/ghidra/util/task/MonitoredRunnable.java#L22

**Reply by topia** [reply]
https://www.oracle.com/technical-resources/articles/java/architect-lambdas-part1.html
> The designers of Java 8 have chosen to give us an annotation, @FunctionalInterface, to serve as a documentation hint that an interface is designed to be used with lambdas, but the compiler does not require this—it determines “functional interfaceness” from the structure of the interface, not from the annotation.

Finally I found that annotation doesn't required at all.

**Reply by astrelsky** [reply]
Huh I wasn't aware of this. Either that or I am confused with something else that had always prevented me from doing it.

**Comment by dev747368** [comment]
In addition to doing your stuff in a Task thread, since you are using a TableModel, you might be able to do a ThreadedTableModel.  There are several examples in the code base, but the main idea is that you do your slow work in your model's doLoad() method.

**Reply by kn0wl3dge** [reply]
Thanks for your reply! I am going to switch to a ThreadedTableModel since it seems to be more adapted to my case.
